rclrs/src/node/primitive_options.rs

```diff
@@ -0,0 +1,253 @@
+/// Each different primitive type may have its own defaults for the overall
+/// quality of service settings, and we cannot know what the default will be
+/// until the `PrimitiveOptions` gets converted into the more specific set of
+/// options. Therefore we store each quality of service field separately so that
+/// we will only override the settings that the user explicitly asked for, and
+/// the rest will be determined by the default settings for each primitive.
+///
+/// [1]: crate::Subscription
+/// [2]: crate::Publisher
+/// [3]: crate::Client
+/// [4]: crate::Service
+#[derive(Debug, Clone, Copy)]
+#[non_exhaustive]
+pub struct PrimitiveOptions<'a> {
+    /// The name that will be used for the primitive
+    pub name: &'a str,
+    /// Override the default [`QoSProfile::history`] for the primitive.
+    pub history: Option<QoSHistoryPolicy>,
+    /// Override the default [`QoSProfile::reliability`] for the primitive.
+    pub reliability: Option<QoSReliabilityPolicy>,
+    /// Override the default [`QoSProfile::durability`] for the primitive.
+    pub durability: Option<QoSDurabilityPolicy>,
+    /// Override the default [`QoSProfile::deadline`] for the primitive.
+    pub deadline: Option<QoSDuration>,
+    /// Override the default [`QoSProfile::lifespan`] for the primitive.
+    pub lifespan: Option<QoSDuration>,
+    /// Override the default [`QoSProfile::liveliness`] for the primitive.
+    pub liveliness: Option<QoSLivelinessPolicy>,
+    /// Override the default [`QoSProfile::liveliness_lease`] for the primitive.
+    pub liveliness_lease: Option<QoSDuration>,
+    /// Override the default [`QoSProfile::avoid_ros_namespace_conventions`] for the primitive.
+    pub avoid_ros_namespace_conventions: Option<bool>,
+}
+
+/// Trait to implicitly convert a compatible object into [`PrimitiveOptions`].
+pub trait IntoPrimitiveOptions<'a>: Sized {
+    /// Convert the object into [`PrimitiveOptions`] with default settings.
+    fn into_primitive_options(self) -> PrimitiveOptions<'a>;
+
+    /// Override all the quality of service settings for the primitive.
+    fn qos(self, profile: QoSProfile) -> PrimitiveOptions<'a> {
+        self.into_primitive_options().history(profile.history)
+    }
+
+    /// Use the default topics quality of service profile.
+    fn topics_qos(self) -> PrimitiveOptions<'a> {
+        self.qos(QoSProfile::topics_default())
+    }
+
+    /// Use the default sensor data quality of service profile.
+    fn sensor_data_qos(self) -> PrimitiveOptions<'a> {
+        self.qos(QoSProfile::sensor_data_default())
+    }
+
+    /// Use the default services quality of service profile.
+    fn services_qos(self) -> PrimitiveOptions<'a> {
+        self.qos(QoSProfile::services_default())
+    }
+
+    /// Use the system-defined default quality of service profile. This profile
+    /// is determined by the underlying RMW implementation, so you cannot rely
+    /// on this profile being consistent or appropriate for your needs.
```

Comment:
Indeed, the language is intentionally very negative, and I would urge us to keep it negative because I believe it should be.

Speaking for myself, in the very early days of using rclcpp I was using `rclcpp::SystemDefaultsQoS` for every primitive that I created with the impression that "system default" meant that for topics it would use the recommended ROS topic default and for services it would use the recommended ROS service defaults. It was some years later that I found out this is not the case at all.. the system default QoS can be anything, and it'll apply the same way to topics, services, and anything else.

I think calling it "system default" was a mistake because of how easily the word "default" in particular will mislead users. But the terminology is now cemented in rclcpp and rclpy, and I think this is one place where it doesn't make sense for rclrs to diverge. So instead I think we should make the docs for it scary enough to get user's attention so that they'll question whether this is something they really want to use. And in virtually all cases this is NOT something they actually want to use.

Reply:
That's an interesting perspective, from my professional usage of ROS 2, we explicitly wanted to use the system default because all of our nodes used the same RMW impl and we wanted to use QoS settings that are outside of the RMW.

What do you think about changing the wording to something like this? 
> topics and services created with this default do not use the recommended ROS defaults, they will instead use the default as defined by the underlying RMW implementation (rmw_fastrtps, rmw_connextdds, etc). These defaults may not always be appropriate in a ROS-compatible system (especially when mixing RMW implementations), so use caution!

Reply:
I've applied your feedback in [5e656c3](https://github.com/ros2-rust/ros2_rust/pull/429/commits/5e656c3f8bd20d9065e59cee3c1f16a876074f5a) with one tweak: I didn't even mention the idea of mixing RMW implementations since that's not officially supported in the first place. Instead I put in a more vague statement of "These defaults may not always be appropriate for every use-case".